middleware/index.js

```diff
@@ -0,0 +1,227 @@
+'use strict';
+var chalk = require('chalk');
+var yeoman = require('yeoman-generator');
+
+var wsModels = require('loopback-workspace').models;
+
+var actions = require('../lib/actions');
+var helpers = require('../lib/helpers');
+var validateName = helpers.validateName;
+var async = require('async');
+
+function listChoices(items) {
+  var choices = [];
+  items.forEach(function(item, index) {
+    choices.push({
+      name: (index + 1) + '. ' + item,
+      value: item
+    });
+  });
+  return choices;
+}
+
+module.exports = yeoman.generators.Base.extend({
+  // NOTE(bajtos)
+  // This generator does not track file changes via yeoman,
+  // as loopback-workspace is editing (modifying) files when
+  // saving project changes.
+
+  loadProject: actions.loadProject,
+
+  constructor: function() {
+    yeoman.generators.Base.apply(this, arguments);
+
+    this.argument('name', {
+      desc: 'Name of the middleware to create.',
+      required: false,
+      type: String
+    });
+  },
+
+  help: function() {
+    return helpers.customHelp(this);
+  },
+
+  loadPhases: function() {
+    var done = this.async();
+    wsModels.Middleware.find(function(err, list) {
+      if (err) {
+        return done(err);
+      }
+
+      var phases = {};
+      list.forEach(function(c) {
+        if (c.phase) {
+          phases[c.phase] = c.phase;
+        }
+      });
+
+      this.phases = Object.keys(phases);
+
+      done();
+    }.bind(this));
+  },
+
+  askForName: function() {
+    var done = this.async();
+
+    var prompts = [
+      {
+        name: 'name',
+        message: 'Enter the middleware name:',
+        default: this.name,
+        validate: function(name) {
+          return name && validateName(name);
+        }
+      }
+    ];
+
+    this.prompt(prompts, function(props) {
+      this.name = props.name;
+      done();
+    }.bind(this));
+
+  },
+
+  askForPhase: function() {
+    var done = this.async();
+
+    var displayName = chalk.yellow(this.name);
+
+    var prompts = [
+      {
+        name: 'phase',
+        message: 'Select the phase for ' + displayName + ':',
+        type: 'list',
+        default: 'routes',
+        choices: listChoices(this.phases.concat(['__other__']))
```

Comment:
FWIW, I find it slightly misleading that this "other" option has a sequence number too. Since this number is the last one, it looks like the new phase will come after "final".

I am proposing to remove the sequence number from the "other" option, i.e.

```
  7. parse
  8. routes
  9. session
> (create a new phase)
```